// @php-wasm/cli-util/index.js
import f from "fs";
import P from "path";
import { toPosixPath as D } from "@php-wasm/util";
import { XMLParser as T, XMLBuilder as $ } from "fast-xml-parser";
import * as v from "jsonc-parser";
const O = "PHPWASMCLI";
async function q(c, a, r) {
  const p = r === "win32" ? (
    // On Windows, creating a 'dir' symlink can require elevated permissions.
    // In this case, let's make junction points because they function like
    // symlinks and do not require elevated permissions.
    "junction"
  ) : "dir";
  f.symlinkSync(c, a, p);
}
async function H(c) {
  try {
    f.lstatSync(c).isSymbolicLink() && f.unlinkSync(c);
  } catch {
  }
}
function A(c, a) {
  return a.filter((r) => {
    const p = P.resolve(r.hostPath), h = P.join(c, P.sep);
    return (
      // If auto-mounting from the current directory,
      // the entire project directory can be mapped.
      p === c || p.startsWith(h)
    );
  });
}
const x = {
  ignoreAttributes: !1,
  attributeNamePrefix: "",
  preserveOrder: !0,
  cdataPropName: "__cdata",
  commentPropName: "__xmlComment",
  allowBooleanAttributes: !0,
  trimValues: !0
}, M = {
  ignoreAttributes: x.ignoreAttributes,
  attributeNamePrefix: x.attributeNamePrefix,
  preserveOrder: x.preserveOrder,
  cdataPropName: x.cdataPropName,
  commentPropName: x.commentPropName,
  suppressBooleanAttributes: !x.allowBooleanAttributes,
  format: !0,
  indentBy: "	"
}, j = {
  allowEmptyContent: !0,
  allowTrailingComma: !0
};
function _(c, a) {
  var E, b, I, F, L, k;
  const { name: r, host: p, port: h, mappings: l, ideKey: t } = a, m = new T(x), g = (() => {
    try {
      return m.parse(c, !0);
    } catch {
      throw new Error("PhpStorm configuration file is not valid XML.");
    }
  })(), u = {
    server: [{}],
    ":@": {
      name: r,
      // NOTE: PhpStorm quirk: Xdebug only works when the full URL (including port)
      // is provided in `host`. The separate `port` field is ignored or misinterpreted,
      // so we rely solely on host: "host:port".
      host: `${p}:${h}`,
      use_path_mappings: "true"
    }
  };
  l && l.length && (u.server[0].path_mappings = l.map((e) => ({
    mapping: [],
    ":@": {
      "local-root": `$PROJECT_DIR$/${D(
        P.relative(a.projectDir, e.hostPath)
      )}`,
      "remote-root": e.vfsPath
    }
  })));
  let n = g == null ? void 0 : g.find((e) => !!(e != null && e.project));
  if (n) {
    const e = (E = n[":@"]) == null ? void 0 : E.version;
    if (e === void 0)
      throw new Error(
        'PhpStorm IDE integration only supports <project version="4"> in workspace.xml, but the <project> configuration has no version number.'
      );
    if (e !== "4")
      throw new Error(
        `PhpStorm IDE integration only supports <project version="4"> in workspace.xml, but we found a <project> configuration with version "${e}".`
      );
  }
  n === void 0 && (n = {
    project: [],
    ":@": { version: "4" }
  }, g.push(n));
  let o = (b = n.project) == null ? void 0 : b.find(
    (e) => {
      var S;
      return !!(e != null && e.component) && ((S = e == null ? void 0 : e[":@"]) == null ? void 0 : S.name) === "PhpServers";
    }
  );
  o === void 0 && (o = {
    component: [],
    ":@": { name: "PhpServers" }
  }, n.project === void 0 && (n.project = []), n.project.push(o));
  let s = (I = o.component) == null ? void 0 : I.find(
    (e) => !!(e != null && e.servers)
  );
  s === void 0 && (s = { servers: [] }, o.component === void 0 && (o.component = []), o.component.push(s));
  const y = (F = s.servers) == null ? void 0 : F.findIndex(
    (e) => {
      var S;
      return !!(e != null && e.server) && ((S = e == null ? void 0 : e[":@"]) == null ? void 0 : S.name) === r;
    }
  );
  (y === void 0 || y < 0) && (s.servers === void 0 && (s.servers = []), s.servers.push(u));
  let d = (L = n.project) == null ? void 0 : L.find(
    (e) => {
      var S;
      return !!(e != null && e.component) && ((S = e == null ? void 0 : e[":@"]) == null ? void 0 : S.name) === "RunManager";
    }
  );
  if (d === void 0 && (d = {
    component: [],
    ":@": { name: "RunManager" }
  }, n.project === void 0 && (n.project = []), n.project.push(d)), (((k = d.component) == null ? void 0 : k.findIndex(
    (e) => {
      var S;
      return !!(e != null && e.configuration) && ((S = e == null ? void 0 : e[":@"]) == null ? void 0 : S.name) === r;
    }
  )) ?? -1) < 0) {
    const e = {
      configuration: [
        {
          method: [],
          ":@": { v: "2" }
        }
      ],
      ":@": {
        name: r,
        type: "PhpRemoteDebugRunConfigurationType",
        factoryName: "PHP Remote Debug",
        filter_connections: "FILTER",
        server_name: r,
        session_id: t
      }
    };
    d.component === void 0 && (d.component = []), d.component.push(e);
  }
  const C = new $(M).build(g);
  try {
    m.parse(C, !0);
  } catch {
    throw new Error(
      "The resulting PhpStorm configuration file is not valid XML."
    );
  }
  return C;
}
function B(c, a) {
  var u, n;
  const { name: r, mappings: p } = a, h = [];
  let l = c, t = v.parseTree(l, h, j);
  if (t === void 0 || h.length)
    throw new Error("VS Code configuration file is not valid JSON.");
  let m = v.findNodeAtLocation(t, ["configurations"]);
  if (m === void 0 || m.children === void 0) {
    const o = v.modify(l, ["configurations"], [], {});
    l = v.applyEdits(l, o), t = v.parseTree(l, [], j), m = v.findNodeAtLocation(t, [
      "configurations"
    ]);
  }
  const g = (u = m == null ? void 0 : m.children) == null ? void 0 : u.findIndex(
    (o) => {
      var s;
      return ((s = v.findNodeAtLocation(o, ["name"])) == null ? void 0 : s.value) === r;
    }
  );
  if (g === void 0 || g < 0) {
    const o = {
      name: r,
      type: "php",
      request: "launch",
      port: 9003
    };
    p && p.length && (o.pathMappings = p.reduce((d, i) => (d[i.vfsPath] = `\${workspaceFolder}/${D(
      P.relative(a.workspaceDir, i.hostPath)
    )}`, d), {}));
    const s = ((n = m == null ? void 0 : m.children) == null ? void 0 : n.length) || 0, y = v.modify(
      l,
      ["configurations", s],
      o,
      {
        formattingOptions: {
          insertSpaces: !0,
          tabSize: 4,
          eol: `
`
        }
      }
    );
    l = N(l, y);
  }
  return l;
}
async function U({
  name: c,
  ides: a,
  host: r,
  port: p,
  cwd: h,
  mounts: l,
  ideKey: t = O
}) {
  const m = l ? A(h, l) : [], g = {};
  if (a.includes("phpstorm")) {
    const u = ".idea/workspace.xml", n = P.join(
      h,
      u
    );
    if (!f.existsSync(n)) {
      if (f.existsSync(P.dirname(n)))
        f.writeFileSync(
          n,
          `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>`
        );
      else if (a.length == 1)
        throw new Error(
          "PhpStorm IDE integration requested, but no '.idea' directory was found in the current working directory."
        );
    }
    if (f.existsSync(n)) {
      const o = f.readFileSync(n, "utf8"), s = _(o, {
        name: c,
        host: r,
        port: p,
        projectDir: h,
        mappings: m,
        ideKey: t
      });
      f.writeFileSync(n, s), g.phpstorm = u;
    }
  }
  if (a.includes("vscode")) {
    const u = ".vscode/launch.json", n = P.join(
      h,
      u
    );
    if (!f.existsSync(n)) {
      if (f.existsSync(P.dirname(n)))
        f.writeFileSync(
          n,
          `{
    "configurations": []
}`
        );
      else if (a.length == 1)
        throw new Error(
          "VS Code IDE integration requested, but no '.vscode' directory was found in the current working directory."
        );
    }
    if (f.existsSync(n)) {
      const o = f.readFileSync(n, "utf-8"), s = B(o, {
        name: c,
        workspaceDir: h,
        mappings: m
      });
      s !== o && (f.writeFileSync(n, s), g.vscode = u);
    }
  }
  return g;
}
async function W(c, a) {
  var h, l, t, m;
  const r = P.join(a, ".idea/workspace.xml");
  if (f.existsSync(r)) {
    const g = f.readFileSync(r, "utf8"), u = new T(x), n = (() => {
      try {
        return u.parse(g, !0);
      } catch {
        throw new Error(
          "PhpStorm configuration file is not valid XML."
        );
      }
    })(), o = n.find(
      (i) => !!(i != null && i.project)
    ), s = (h = o == null ? void 0 : o.project) == null ? void 0 : h.find(
      (i) => {
        var w;
        return !!(i != null && i.component) && ((w = i == null ? void 0 : i[":@"]) == null ? void 0 : w.name) === "PhpServers";
      }
    ), y = (l = s == null ? void 0 : s.component) == null ? void 0 : l.find(
      (i) => !!(i != null && i.servers)
    ), d = (t = y == null ? void 0 : y.servers) == null ? void 0 : t.findIndex(
      (i) => {
        var w;
        return !!(i != null && i.server) && ((w = i == null ? void 0 : i[":@"]) == null ? void 0 : w.name) === c;
      }
    );
    if (d !== void 0 && d >= 0) {
      y.servers.splice(d, 1);
      const w = new $(M).build(n);
      try {
        u.parse(w, !0);
      } catch {
        throw new Error(
          "The resulting PhpStorm configuration file is not valid XML."
        );
      }
      w === `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
	<component name="PhpServers">
		<servers></servers>
	</component>
</project>` ? f.unlinkSync(r) : f.writeFileSync(r, w);
    }
  }
  const p = P.join(a, ".vscode/launch.json");
  if (f.existsSync(p)) {
    const g = [], u = f.readFileSync(p, "utf-8"), n = v.parseTree(u, g, j);
    if (n === void 0 || g.length)
      throw new Error("VS Code configuration file is not valid JSON.");
    const o = v.findNodeAtLocation(n, [
      "configurations"
    ]), s = (m = o == null ? void 0 : o.children) == null ? void 0 : m.findIndex(
      (y) => {
        var d;
        return ((d = v.findNodeAtLocation(y, ["name"])) == null ? void 0 : d.value) === c;
      }
    );
    if (s !== void 0 && s >= 0) {
      const y = v.modify(
        u,
        ["configurations", s],
        void 0,
        {
          formattingOptions: {
            insertSpaces: !0,
            tabSize: 4,
            eol: `
`
          }
        }
      ), d = N(u, y);
      d === `{
    "configurations": []
}` ? f.unlinkSync(p) : f.writeFileSync(p, d);
    }
  }
}
function z({
  cwd: c,
  mounts: a,
  pathSkippings: r
}) {
  return { pathMappings: c && a ? A(c, a) : void 0, pathSkippings: r };
}
function N(c, a) {
  const r = [], p = v.applyEdits(c, a);
  if (r.length = 0, v.parseTree(p, r, j), r.length) {
    const h = r.map((t) => ({
      message: v.printParseErrorCode(t.error),
      offset: t.offset,
      length: t.length,
      fragment: p.slice(
        Math.max(0, t.offset - 20),
        Math.min(p.length, t.offset + t.length + 10)
      )
    })).map(
      (t) => `${t.message} at ${t.offset}:${t.length} (${t.fragment})`
    ), l = a.map(
      (t) => `At ${t.offset}:${t.length} - (${t.content})`
    );
    throw new Error(
      `VS Code configuration file (.vscode/launch.json) is not valid a JSONC after CLI modifications. This is likely a CLI bug. Please report it at https://github.com/WordPress/wordpress-playground/issues and include the contents of your ".vscode/launch.json" file. 

 Applied edits: ${l.join(
        `
`
      )}

 The errors are: ${h.join(`
`)}`
    );
  }
  return p;
}
export {
  O as DEFAULT_IDE_KEY,
  U as addXdebugIDEConfig,
  W as clearXdebugIDEConfig,
  q as createTempDirSymlink,
  z as makeXdebugConfig,
  H as removeTempDirSymlink,
  _ as updatePhpStormConfig,
  B as updateVSCodeConfig
};
