// @php-wasm/web-service-worker/index.js
import { isURLScoped as b, setURLScope as C, getURLScope as w } from "@php-wasm/scopes";
const E = 25e3;
let L = 0;
function H(e, t, ...r) {
  const s = m();
  return e.postMessage(
    {
      ...t,
      requestId: s
    },
    ...r
  ), s;
}
function m() {
  return ++L;
}
function S(e, t, r = E) {
  return new Promise((s, n) => {
    const o = (c) => {
      c.data.type === "response" && c.data.requestId === t && (e.removeEventListener("message", o), clearTimeout(a), s(c.data.response));
    }, a = setTimeout(() => {
      n(new Error("Request timed out")), e.removeEventListener("message", o);
    }, r);
    e.addEventListener("message", o);
  });
}
function I(e, t) {
  return {
    type: "response",
    requestId: e,
    response: t
  };
}
async function O(e) {
  let t = new URL(e.request.url);
  if (!b(t))
    try {
      const i = new URL(e.request.referrer);
      t = C(t, w(i));
    } catch {
    }
  const r = e.request.headers.get("content-type"), s = e.request.method === "POST" ? new Uint8Array(await e.request.clone().arrayBuffer()) : void 0, n = {};
  for (const i of e.request.headers.entries())
    n[i[0]] = i[1];
  let o;
  try {
    const i = {
      method: "request",
      args: [
        {
          body: s,
          url: t.toString(),
          method: e.request.method,
          headers: {
            ...n,
            Host: t.host,
            // Safari and Firefox don't make the User-Agent header
            // available in the fetch event. Let's add it manually:
            "User-agent": self.navigator.userAgent,
            "Content-type": r
          }
        }
      ]
    }, d = w(t);
    if (d === null)
      throw new Error(
        `The URL ${t.toString()} is not scoped. This should not happen.`
      );
    const f = await U(i, d);
    if (o = await S(self, f), delete o.headers["x-frame-options"], o.headers["content-security-policy"]) {
      const l = o.headers["content-security-policy"].map(
        (u) => x(
          "frame-ancestors",
          u
        )
      ).filter((u) => u.trim().length > 0);
      l.length > 0 ? o.headers["content-security-policy"] = l : delete o.headers["content-security-policy"];
    }
  } catch (i) {
    throw console.error(i, { url: t.toString() }), i;
  }
  if (o.httpStatusCode >= 300 && o.httpStatusCode <= 399 && o.headers.location)
    return Response.redirect(
      new URL(o.headers.location[0], t.toString()),
      o.httpStatusCode
    );
  const c = [101, 103, 204, 205, 304].includes(
    o.httpStatusCode
  ) ? null : o.bytes;
  return new Response(c, {
    headers: o.headers,
    status: o.httpStatusCode
  });
}
async function U(e, t) {
  const r = m();
  for (const s of await self.clients.matchAll({
    // Sometimes the client that triggered the current fetch()
    // event is considered uncontrolled in Google Chrome. This
    // only happens on the first few fetches() after the initial
    // registration of the service worker.
    includeUncontrolled: !0
  }))
    s.postMessage({
      ...e,
      /**
       * Attach the scope with a URL starting with `/scope:` to this message.
       *
       * We need this mechanics because this worker broadcasts
       * events to all the listeners across all browser tabs. Scopes
       * helps WASM workers ignore requests meant for other WASM workers.
       */
      scope: t,
      requestId: r
    });
  return r;
}
async function p(e, t) {
  let r;
  return ["GET", "HEAD"].includes(e.method) || "body" in t ? r = void 0 : !e.bodyUsed && e.body ? r = e.body : r = await e.arrayBuffer(), new Request(t.url || e.url, {
    body: r,
    method: e.method,
    headers: e.headers,
    referrer: e.referrer,
    referrerPolicy: e.referrerPolicy,
    mode: e.mode === "navigate" ? "same-origin" : e.mode,
    credentials: e.credentials,
    cache: e.cache,
    redirect: e.redirect,
    integrity: e.integrity,
    ...r instanceof ReadableStream && { duplex: "half" },
    ...t
  });
}
async function T(e) {
  if (!e.body)
    return [e, e];
  const [t, r] = e.body.tee();
  return [
    await p(e, { body: t, duplex: "half" }),
    await p(e, { body: r, duplex: "half" })
  ];
}
function D(e) {
  const t = {};
  return e.headers.forEach((r, s) => {
    t[s] = r;
  }), t;
}
function x(e, t) {
  const r = /^[\u{9}\u{A}\u{C}\u{D}\u{20}]+/u, s = /[\u{9}\u{A}\u{C}\u{D}\u{20}]+$/u, n = /[\u{9}\u{A}\u{C}\u{D}\u{20}]/u;
  return t.split(";").filter((o) => {
    const a = o.replace(r, "").replace(s, ""), [c] = a.split(
      n,
      // The directive name is the first token.
      1
    );
    return c.toLowerCase() !== e.toLowerCase();
  }).join(";");
}
class P extends Error {
  constructor(t, r, s) {
    super(
      `Could not fetch ${t} – your network appears to be blocking this request (HTTP ${r}). This often happens on school, university, or corporate networks. Try switching to a different network or using a VPN.`
    ), this.name = "FirewallInterferenceError", this.url = t, this.status = r, this.statusText = s;
  }
}
const A = "X-Playground-Cors-Proxy";
async function v(e, t, r, s) {
  var f;
  let n = typeof e == "string" ? new Request(e, t) : e;
  const o = s ? new URL(s) : null;
  let a = o ? new URL(n.url, o) : new URL(n.url);
  if (a.hostname === "localhost" || a.hostname === "127.0.0.1" || a.hostname === "[::1]" || a.hostname === "::1")
    return await h(n);
  if (a.protocol === "http:") {
    a.protocol = "https:";
    const l = a.toString();
    n = await p(n, { url: l }), a = new URL(l);
  }
  if (!r)
    return await h(n);
  if (o && a.protocol === o.protocol && a.hostname === o.hostname && a.port === o.port && a.pathname.startsWith(o.pathname))
    return await h(n);
  const [i, d] = await T(n);
  try {
    return await h(i);
  } catch {
    const u = ((f = new Headers(d.headers).get("x-cors-proxy-allowed-request-headers")) == null ? void 0 : f.split(",")) || [], R = u.includes("authorization") || u.includes("cookie"), g = await p(d, {
      url: `${r}${n.url}`,
      ...R && { credentials: "include" }
    }), y = await h(g, t);
    if (!y.headers.has(A))
      throw new P(
        n.url,
        y.status,
        y.statusText
      );
    return y;
  }
}
async function h(e, t) {
  let r = t ? new Request(e, t) : e;
  if (new URL(r.url).protocol === "http:" && r.body) {
    const s = await new Response(r.body).arrayBuffer();
    r = await p(r, {
      body: s
    });
  }
  return fetch(r);
}
export {
  P as FirewallInterferenceError,
  S as awaitReply,
  U as broadcastMessageExpectReply,
  p as cloneRequest,
  O as convertFetchEventToPHPRequest,
  v as fetchWithCorsProxy,
  m as getNextRequestId,
  D as getRequestHeaders,
  H as postMessageExpectReply,
  x as removeContentSecurityPolicyDirective,
  I as responseTo,
  T as teeRequest
};
